feat(maker): Monitor and expose status of services connected

klochowicz opened the pull request:
Monitor the following services:
    - coordinator
    - orderbook (we don't subscribe to orderbook feed yet)

in order to derive health status of the maker.

Health status is exposed via the HTTP API as well as reported as
dedicated prometheus metrics.

- [ ] Subscribe to orderbook websocket stream to be able to detect if it is down
- [ ] Subscribe to bitmex client stream to be able to spot outdated information

the TODO listed above will be wired only when bitmex and orderbook streams will be integrated into maker, and as such are *not* a dependency of this PR - we can wire them in the corresponding PRs or in a follow-up.

klochowicz commented:
I'm not sure whether we actually need the orderbook status (we don't seem to rely on the websocket stream at all so far? 🤔), and bitmex status can be added when we actually start hedging (it was not mentioned in the ticket, but I think it's an essential service for the maker to be monitored).

Restioson commented on crates/tests-e2e/src/maker.rs:
```diff
@@ -29,8 +29,10 @@ impl Maker {
+    /// it depends on
     pub async fn is_running(&self) -> bool {
-        self.get("/").await.is_ok()
+        self.get("/health").await.is_ok()
```
Do we want to interpret the healthcheck result here? Or is it always okay if it returns 200?

klochowicz replied:
added interpreting of the result.

Restioson commented on maker/src/health.rs:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    // TODO: Any ideas of the most useful way of sending this information?
+    // - perhaps a serialized struct (json) with the status of each service?
```
Even just a struct here could make sense - it could then later be serialized if needed, or logged just like that with `{:?}`

bonomat replied:
Same thought: why are you not returning a struct serialized to json?

klochowicz replied:
changed to return serialised json.

bonomat commented on maker/src/routes.rs:
```diff
@@ -52,6 +56,7 @@ pub fn router(
         .route("/api/pay-invoice/:invoice", post(pay_invoice))
         .route("/api/sync-on-chain", post(sync_on_chain))
         .route("/metrics", get(get_metrics))
+        .route("/health", get(get_health))
```
For whom is this new endpoint?

klochowicz replied:
it's useful in development, testing, and manual checking if we want to quickly see the state.

bonomat commented on maker/src/health.rs:
```diff
@@ -0,0 +1,120 @@
+    pub fn get_orderbook_status(&self) -> ServiceStatus {
+        *self.orderbook_rx.borrow()
+    }
+    // TODO: Add bitmex status? would anything else be useful here?
```
Should this rather be captured in a ticket?

klochowicz replied:
this was not in the original scope of the ticket, so I have left it as a comment first - wasn't sure whether we wanted it! 
it seems it can be addressed rather quickly, so I'll add support for it in this PR.

bonomat commented on maker/src/health.rs:
```diff
@@ -0,0 +1,120 @@
+
+        if self.get_coordinator_status() == ServiceStatus::Online
+        // TODO: Uncomment this line when we retrieve the orderbook status
+        // && (self.get_orderbook_status() == ServiceStatus::Online)
```
Does this work already?

klochowicz replied:
if by "orderbook" we mean websocket stream that other people look at orders, then not, we are still not subscribing to it from the maker. I was hoping that other work done for this milestone would encompass this, but nothing got still merged.

@luckysori do I remember correctly that you were adding the orderbook websocket stream?

klochowicz replied:
edit: I have removed the TODO from this place and hard-coded `Online` status for the orderbook where it should be plugged in (there's another TODO there anyway).

bonomat commented on maker/src/metrics.rs:
```diff
@@ -58,12 +69,42 @@ pub fn init_meter() -> PrometheusExporter {
+/// Updates the health metric given a service status
+fn update_health_metric(cx: &Context, service_status: ServiceStatus, gauge: &ObservableGauge<u64>) {
+    match service_status {
+        ServiceStatus::Unknown => {
+            gauge.observe(cx, 0, &[KeyValue::new("status", "unknown")]);
+        }
+        ServiceStatus::Online => {
+            gauge.observe(cx, 1, &[KeyValue::new("status", "online")]);
+        }
+        ServiceStatus::Offline => {
+            gauge.observe(cx, 0, &[KeyValue::new("status", "offline")]);
+        }
```
🚨 I don't think this works like you want: 

Our grafana agent will poll this metric ever X seconds but it will not reset an old value it has seen before, which means, with this setup, we might have a service which is online and offline at the same time, i.e. 

at 10:00:00 service was offline: we create the metric: 
`coordinator_status{status=offline}=0`

at 11:00:00 the service is online, we create the metric
`coordinator_status{status=online}=1`

This means, at 11:00:00, in our grafana dashboard we will have both values set: 
```
coordinator_status{status=offline}=0
coordinator_status{status=online}=1
```

Because `coordinator_status{status=offline}` and `coordinator_status{status=online}` are two different metrics.

Instead, you need to use the same metric to hold different states, e.g. 

- `coordinator_status=0` = offline
- `coordinator_status=1` = online
- `coordinator_status=2` = unknown --> with this we could then indicate different error codes.

klochowicz replied:
indeed, I had a wrong mental model how attributes for metrics work. Thank you for explanation, I've fixed it now.


cool idea with mapping `2 == unknown` .

bonomat reviewed (changes requested):
There are a few comments/todos in the code and I believe the usage of the metrics is not correct.
Let me know if my comment is not clear.

klochowicz commented:
> There are a few comments/todos in the code and I believe the usage of the metrics is not correct.
> Let me know if my comment is not clear.

I believe I fixed the issue.
the only `todo` left in code is about us not using the orderbook websocket at all yet, so there's not much I can do about it. I've added bitmex stream.

bonomat reviewed (approved):
LGTM

bonomat commented:
> > There are a few comments/todos in the code and I believe the usage of the metrics is not correct.
> > Let me know if my comment is not clear.
> 
> I believe I fixed the issue. the only `todo` left in code is about us not using the orderbook websocket at all yet, so there's not much I can do about it. I've added bitmex stream monitoring.

Thanks for addressing.
